### PR TITLE
Adding support for --log-level and concurrent S3 get

### PR DESCRIPTION
**Pull request by saurabh-hirani**
This PR adds the following features:

1. Support for --log-level flag on command line to control the logging.

**Demo**

```
go run main.go aws --resource-type s3 --log-level debug
```

or 

```
LOG_LEVEL=debug go run main.go aws --resource-type s3
```

Valid log types are mentioned in the corresponding readme change. 

2. Optimizing the `s3.getAllS3Buckets` function to get S3 bucket properties (region,tags) concurrently as opposed to getting for each one - this was added because running S3 bucket deletion for an account > 50 buckets was too slow and was taking around 1 sec per bucket to get its properties - which would lead to linear increase as per the number of buckets. 

`s3.getAllS3Buckets` now supports batchSize which is as per `aws.MaxConcurrentGetSize` in `s3_types.go` to get X buckets data concurrently (default value: 100)

**Demo**:

1. Before adding concurrent gets:

```
go run main.go aws --resource-type s3

INFO[2020-04-04T21:10:19+05:30] The following resources types will be nuked:
INFO[2020-04-04T21:10:19+05:30] - s3
INFO[2020-04-04T21:10:21+05:30] Retrieving active AWS resources in [eu-north-1, ap-south-1, eu-west-3, eu-west-2, eu-west-1, ap-northeast-2, ap-northeast-1, sa-east-1, ca-central-1, ap-southeast-1, ap-southeast-2, eu-central-1, us-east-1, us-east-2, us-west-1, us-west-2]
INFO[2020-04-04T21:10:21+05:30] Checking region [1/16]: eu-north-1
DEBU[2020-04-04T21:10:23+05:30] Checking - Bucket bucket-1
DEBU[2020-04-04T21:10:26+05:30] Checking - Bucket bucket-2
DEBU[2020-04-04T21:10:33+05:30] Checking - Bucket bucket-3
DEBU[2020-04-04T21:10:39+05:30] Checking - Bucket bucket-4
```

The "Checking -" log - gets bucket region, checks for valid tags and decides if the bucket should be deleted or not. This serial check slows down for accounts having large number of buckets i.e at least n seconds for n buckets.

2. After adding concurrent gets:

```
go run main.go aws --resource-type s3 --log-level debug
INFO[2020-04-04T21:13:50+05:30] The following resources types will be nuked:
INFO[2020-04-04T21:13:50+05:30] - s3
INFO[2020-04-04T21:13:52+05:30] Retrieving active AWS resources in [eu-north-1, ap-south-1, eu-west-3, eu-west-2, eu-west-1, ap-northeast-2, ap-northeast-1, sa-east-1, ca-central-1, ap-southeast-1, ap-southeast-2, eu-central-1, us-east-1, us-east-2, us-west-1, us-west-2]
INFO[2020-04-04T21:13:52+05:30] Checking region [1/16]: eu-north-1
DEBU[2020-04-04T21:13:57+05:30] S3 - creating session - region eu-north-1
DEBU[2020-04-04T21:13:57+05:30] S3 - creating session - region ap-south-1
DEBU[2020-04-04T21:13:57+05:30] S3 - creating session - region eu-west-3
DEBU[2020-04-04T21:13:57+05:30] S3 - creating session - region eu-west-2
DEBU[2020-04-04T21:13:57+05:30] S3 - creating session - region eu-west-1
DEBU[2020-04-04T21:13:57+05:30] S3 - creating session - region ap-northeast-2
DEBU[2020-04-04T21:13:57+05:30] S3 - creating session - region ap-northeast-1
.
.
.
INFO[2020-04-04T21:13:57+05:30] Getting - 1-88 buckets of batch 1/1
```

This run creates S3 bucket sessions in advance (required as each bucket needs its region's session object and we have to create them per region sessions in advance to avoid creating them multiple times).

The log **Getting 1-88 buckets of batch 1/1** gets 88 buckets data in parallel subject to limit imposed by `MaxConcurrentGetSize` in `s3_types.go` and completes in 5-8 seconds as opposed to minimum of 88 seconds for 88 buckets when get operation is serial.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -124,6 +124,16 @@ If you want to check what resources are going to be targeted without actually te
 cloud-nuke aws --resource-type ec2 --dry-run
 ```
 
+### Log level
+
+You can set the log level by specifying the `--log-level` flag as per [logrus](https://github.com/sirupsen/logrus) log levels:
```
Comment (brikis98) [inline README.md:129]: Mention env var too?

File: aws/s3.go [hunk 2]
```diff
@@ -88,55 +97,131 @@ func getAllS3Buckets(awsSession *session.Session, excludeAfter time.Time, bucket
...
+	}
+
+	if batchSize <= 0 {
+		return nil, fmt.Errorf("Invalid batchsize - %d - should be > 0", batchSize)
+	}
+
+	totalBatches := totalBuckets / batchSize
+	if totalBuckets%batchSize != 0 {
+		totalBatches = totalBatches + 1
+	}
```
Comment (brikis98) [inline aws/s3.go:123]: Use `math.Ceil(totalBuckets / batchSize)` instead?

File: aws/s3.go [hunk 3]
```diff
@@ -88,55 +97,131 @@ func getAllS3Buckets(awsSession *session.Session, excludeAfter time.Time, bucket
...
+	batchCount := 1
+
+	// Batch the get operation
+	for i := 0; i < totalBuckets; i += batchSize {
+		j := i + batchSize
```
Comment (brikis98) [inline aws/s3.go:128]: `i` and `j` are hard to follow here. Perhaps `batchStart` and `batchEnd`?

File: aws/s3.go [hunk 4]
```diff
@@ -88,55 +97,131 @@ func getAllS3Buckets(awsSession *session.Session, excludeAfter time.Time, bucket
...
+		if j > totalBuckets {
+			j = totalBuckets
+		}
```
Comment (brikis98) [inline aws/s3.go:131]:
```go
j := math.Min(i + batchSize, totalBuckets)
```

File: aws/s3.go [hunk 5]
```diff
@@ -88,55 +97,131 @@ func getAllS3Buckets(awsSession *session.Session, excludeAfter time.Time, bucket
...
 
-		bucketRegion, err := getS3BucketRegion(svc, bucket.Name)
-		if err != nil {
-			logging.Logger.Warnf("Skipping - Bucket %s - Failed to get bucket location.", *bucket.Name)
-			continue
-		}
+		go func() {
+			wg.Wait()
+			close(bucketCh)
+		}()
```
Comment (brikis98) [inline aws/s3.go:158]: Why do this in a separate goroutine? Don't you usually call `wg.Wait()` in the original calling goroutine so that routine waits for all the others to finish?

Comment (brikis98) [reply aws/s3.go:158]: I guess you're reading from the channel below, but is there any need to _read_ from the channel concurrently? Seems like the code would be easier to reason about if you structure it to:

1. Start N goroutines to fetch bucket information concurrently
1. When all N goroutines are done, read the data they fetched from the channel

File: aws/s3.go [hunk 6]
```diff
@@ -88,55 +97,131 @@ func getAllS3Buckets(awsSession *session.Session, excludeAfter time.Time, bucket
...
 
-		// Create session for current bucket region as bucket location's session is required for getting bucket tags.
-		if _, ok := regionClients[bucketRegion]; !ok {
-			logging.Logger.Debugf("Creating session - Bucket %s - region %s", *bucket.Name, bucketRegion)
-			awsSession, err := session.NewSession(&awsgo.Config{
-				Region: awsgo.String(bucketRegion)},
-			)
-			if err != nil {
-				return bucketNamesPerRegion, errors.WithStackTrace(err)
+		for b := range bucketCh {
```
Comment (brikis98) [inline aws/s3.go:160]: This function has gotten really, really long. Consider breaking into separate pieces: e.g., one that fetches the data concurrently; one that processes it; one that orchestrates the whole thing.

File: aws/s3.go [hunk 7]
```diff
@@ -88,55 +97,131 @@ func getAllS3Buckets(awsSession *session.Session, excludeAfter time.Time, bucket
...
 
-		bucketNamesPerRegion[bucketRegion] = append(bucketNamesPerRegion[bucketRegion], bucket.Name)
+		batchCount++
 	}
 
 	return bucketNamesPerRegion, nil
 }
 
+func getBucketInfo(svc *s3.S3, bucket *s3.Bucket, regionClients map[string]*s3.S3, bucketCh chan<- *S3Bucket) error {
+	var b S3Bucket
```
Comment (brikis98) [inline aws/s3.go:190]: Please minimize usage of single-variable names. E.g., `bucketData` instead of `b`.

File: aws/s3.go (hunk 6 above)
Comment (brikis98) [inline aws/s3.go:160]: Are you checking for errors in the bucket structs being sent on that channel??

File: aws/s3.go [hunk 8]
```diff
@@ -56,26 +57,34 @@ func getS3BucketTags(svc *s3.S3, bucketName *string) ([]map[string]string, error
...
-	}
-	return true, ""
+	return true
+}
+
+// S3Bucket - represents S3 bucket
+type S3Bucket struct {
+	Name           *string
+	CreationDate   *time.Time
+	Region         *string
```
Comment (brikis98) [inline aws/s3.go:78]: Why are these all pointers?

File: aws/s3.go [hunk 9]
```diff
@@ -88,55 +97,131 @@ func getAllS3Buckets(awsSession *session.Session, excludeAfter time.Time, bucket
...
 
+func getBucketInfo(svc *s3.S3, bucket *s3.Bucket, regionClients map[string]*s3.S3, bucketCh chan<- *S3Bucket) error {
+	var b S3Bucket
+	b.Name = bucket.Name
+	b.CreationDate = bucket.CreationDate
+
+	bucketRegion, err := getS3BucketRegion(svc, b.Name)
+	if err != nil {
+		b.Error = err
+		return err
```
Comment (brikis98) [inline aws/s3.go:197]: Why return anything from this function? It looks like you're sending the errors back via a channel (in the struct), so as long as you check for those, it seems like this function shouldn't return anything...

File: aws/s3_types.go [hunk 10]
```diff
@@ -23,6 +23,12 @@ func (bucket S3Buckets) MaxBatchSize() int {
 	return 500
 }
 
+// MaxConcurrentGetSize decides how many S3 buckets to fetch in one call.
+func (bucket S3Buckets) MaxConcurrentGetSize() int {
+	// To speed up bucket fetch part.
+	return 100
```
Comment (brikis98) [inline aws/s3_types.go:29]: Any particular reason for this number? Is there some AWS limit/throttle that drives this?

File: README.md (hunk 1 above)
Comment (saurabh-hirani) [reply README.md:129]: Done

File: aws/s3.go (hunk 2 above)
Comment (saurabh-hirani) [reply aws/s3.go]: Done

File: aws/s3.go (hunk 4 above)
Comment (saurabh-hirani) [reply aws/s3.go]: Done

File: aws/s3.go (hunk 5 above)
Comment (saurabh-hirani) [reply aws/s3.go]: Reading from the channel with data as it comes helps in showing messages earlier for skipped buckets when debug mode is on - as opposed to logging everything when everything is done.

File: aws/s3.go (hunk 7 above)
Comment (saurabh-hirani) [reply aws/s3.go]: Done

File: aws/s3.go (hunk 6 above)
Comment (saurabh-hirani) [reply aws/s3.go]: Added the check.

File: aws/s3.go (hunk 8 above)
Comment (saurabh-hirani) [reply aws/s3.go]: They don't need to be. Fixed.

File: aws/s3.go (hunk 9 above)
Comment (saurabh-hirani) [reply aws/s3.go]: True. Fixed.

File: aws/s3_types.go (hunk 10 above)
Comment (saurabh-hirani) [reply aws/s3_types.go:29]: No specific reason - I didn't want to execute too many goroutines e.g. if 5000 buckets are there - I didn't want to spawn 5000 goroutines and I did not know a right number - added 100 as a round figure.

**Comment by saurabh-hirani** [comment]
@brikis98 - thanks for the feedback - updated and replied to comments. Let me know if it looks good - will squash the commits.

File: aws/s3.go (hunk 3 above)
Comment (saurabh-hirani) [reply aws/s3.go]: Yes. Done.

File: aws/s3.go [hunk 11]
```diff
@@ -2,22 +2,25 @@ package aws
...
 	"github.com/gruntwork-io/cloud-nuke/logging"
 	"github.com/gruntwork-io/gruntwork-cli/errors"
 )
 
 // getS3BucketRegion returns S3 Bucket region.
-func getS3BucketRegion(svc *s3.S3, bucketName *string) (string, error) {
+func getS3BucketRegion(svc *s3.S3, bucketName string) (string, error) {
 	input := &s3.GetBucketLocationInput{
-		Bucket: bucketName,
+		Bucket: &bucketName,
```
Comment (brikis98) [inline aws/s3.go:23]: NIT: The idiom is to use `aws.String()` for converting args for the AWS SDK into pointers. Applies here and below.

File: aws/s3.go [hunk 12]
```diff
@@ -56,87 +65,197 @@ func getS3BucketTags(svc *s3.S3, bucketName *string) ([]map[string]string, error
...
+	}
+
+	go func() {
+		wg.Wait()
+		close(bucketCh)
+	}()
+
+	for bucketData := range bucketCh {
+		if bucketData.Error != nil {
+			logging.Logger.Warnf("Skipping - Bucket %s - region - %s - error: %s", bucketData.Name, bucketData.Region, bucketData.Error)
```
Comment (brikis98) [inline aws/s3.go:190]: Should errors be skipped? Or cause the entire nuke operation to fail due to an error?

File: aws/s3.go [hunk 13]
```diff
@@ -56,87 +65,197 @@ func getS3BucketTags(svc *s3.S3, bucketName *string) ([]map[string]string, error
...
+	return bucketNamesPerRegion, nil
+}
 
-		if _, ok := bucketNamesPerRegion[bucketRegion]; !ok {
-			bucketNamesPerRegion[bucketRegion] = []*string{}
+// getBucketInfo populates the local S3Bucket struct for the passed AWS bucket
+func getBucketInfo(svc *s3.S3, bucket *s3.Bucket, excludeAfter time.Time, regionClients map[string]*s3.S3, bucketCh chan<- *S3Bucket) {
+	var bucketData S3Bucket
+	bucketData.Name = *bucket.Name
+	bucketData.CreationDate = *bucket.CreationDate
```
Comment (brikis98) [inline aws/s3.go:209]: NITT: *Some* AWS params can be nil pointers. It's usually safer to use functions like `aws.StringValue` and `aws.TimeValue` to dereference these.

File: aws/s3.go [hunk 14]
```diff
@@ -56,87 +65,197 @@ func getS3BucketTags(svc *s3.S3, bucketName *string) ([]map[string]string, error
...
+	// Check if the bucket is in target region
+	matchedRegion := false
+	for region := range regionClients {
+		if region == bucketData.Region {
+			matchedRegion = true
+			break
 		}
+	}
+	if !matchedRegion {
+		bucketData.InvalidReason = "Not in target region"
```
Comment (brikis98) [inline aws/s3.go:228]: How would we have gotten a bucket not in one of those regions? Is this just an extra safety check?

File: aws/s3_test.go [hunk 15]
```diff
@@ -60,21 +76,22 @@ func (b TestS3Bucket) create(svc *s3.S3) error {
 
 	// Add default tag for testing
 	var awsTagSet []*s3.Tag
-	awsTagSet = append(awsTagSet, &s3.Tag{Key: aws.String("cloud-nuke-test"), Value: aws.String("true")})
```
Comment (brikis98) [inline aws/s3_test.go:63]: Why remove this?

File: aws/s3.go (hunk 5 above)
Comment (brikis98) [reply aws/s3.go]: Roger. Please add that as a comment!

File: aws/s3.go (hunk 11 above)
Comment (saurabh-hirani) [reply aws/s3.go]: Done.

File: aws/s3.go (hunk 12 above)
Comment (saurabh-hirani) [reply aws/s3.go:190]: As per earlier conversation - https://github.com/gruntwork-io/cloud-nuke/pull/96#discussion_r386855853 - the aim was to delete as many buckets as we can and then return a deleteCount which calls out if the deletion was 100% successfull or not https://github.com/gruntwork-io/cloud-nuke/blob/f579288286105f472751718c257638a5329834a6/aws/s3_types.go#L48

File: aws/s3.go (hunk 13 above)
Comment (saurabh-hirani) [reply aws/s3.go]: Thanks for catching these. Done.

File: aws/s3.go (hunk 14 above)
Comment (saurabh-hirani) [reply aws/s3.go:228]: For S3 - you have to list out all buckets and find each bucket's individual region. Earlier we were listing all buckets, finding all regions and returning it. This wasn't required because if there is a target region specified by end user (or all regions if not specified) - and if a bucket is not in the target region - we will be skipping it later - better to skip it at the start.

File: aws/s3_test.go (hunk 15 above)
Comment (saurabh-hirani) [reply aws/s3_test.go:63]: This prevented testing the situation of deleting an untagged bucket. The earier PR was so focused on deletion through tags, that the untagged scenario went untests. Added that in this PR + for test cases we are passing "cloud-nuke-test" when testing for tagged scenarios.

File: aws/s3.go (hunk 5 above)
Comment (saurabh-hirani) [reply aws/s3.go]: Done.

**Comment by brikis98** [comment]
Tests passed! Merging now.

**Comment by brikis98** [comment]
https://github.com/gruntwork-io/cloud-nuke/releases/tag/v0.1.18

**Comment by saurabh-hirani** [comment]
Thanks @brikis98 for the feedback and code review!
